Add tests ensuring the tracing module sets and cleans up metadata properly

oleiade opened the pull request:
This Pull Request addresses #2889, and adds tests ensuring that the tracing module does add a `trace_id` metadata entry when performing a request and cleans it up afterward.

To make it possible, I had to wrap a bunch of existing behaviors throughout the module in interfaces. That way I could easily swap them with mocked/controllable behaviors: trace id generation and randomness generation.

Let me know what you think 🤝 

closes #2889

mstoykov commented on js/modules/k6/experimental/tracing/trace_id.go:
```diff
@@ -45,6 +46,23 @@ type TraceID struct {
+		Prefix:     prefix,
+		Code:       code,
+		Time:       t,
+		randSource: rand.Reader,
```
off-topic: Does the source need to be crypto secure?

oleiade replied:
As far as I know, it doesn't need to. I am switching to `math/rand.Rand` 👍🏻

mstoykov reviewed (dismissed):
LGTM!

I was thinking more of js code test that has 
```javascript
callToTestMetadata()
http.post(...)
callToTestMetadata() 
```

maybe something like what I do in the async group [code](https://github.com/grafana/k6/blob/8101130bf73b285f5305c450b64ad9e017002a3c/js/modules/k6/asyncGroup_test.go#L126) but with a bertter name, The script that is run is in test case sabove the line I linked.

Goja's author also does something similar in https://github.com/dop251/goja/blob/8a0df7296ebdd6b66ab2bb8230f9744dc8824825/func_test.go#L212

oleiade commented:
@mstoykov that's another clever way of doing it indeed 🤝 When looking at how to access the metadata before the callback is called, I didn't think of doing it from the JS indeed.

In this case, I find the unit test ends up being "lower-context", forcing me to make the code a bit more flexible by using interfaces. If you have a strong opinion towards making it in JS, I could give it a shot too. Otherwise, I would keep it as is as unit tests and the JS approach in mind for the next time. Let me know 😸

oleiade commented on js/modules/k6/experimental/tracing/producer.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+// Producer is an interface for producing trace IDs.
+type Producer interface {
+	Produce() TraceID
+}
+
+// DefaultProducer is a default implementation of the Producer interface.
+type DefaultProducer struct{}
```
Nope, you're right 👍🏻

codebien reviewed (approved):
LGTM

codebien commented:
I have re-run the test-tip check to verify it is just one of the flaky.

mstoykov reviewed (changes requested):
Having now gone through the code one more time and through https://github.com/grafana/k6/pull/2886 it seems like the new Producer only purpose is to make the current test possible.

There is also no "pruducer" abstraction or terminology within the tracing vocabulary as well. 

Given just how much code this adds and that I feel like it makes everything *harder* and the fact this all can be tested with just a single defintion of a helper js function as I said before - I would prefer we go that route and this  PR to only add tests instead of also refactor.

Additionally this will hopefully make this PR have less conflicts with the sampling one

oleiade commented:
@mstoykov Although I disagree with the perceived complexity introduced by the approach in this PR, I see your point, and I'm indeed open to doing it in JS.

After experimenting with your examples and experimenting with your approach, I want to make one of my findings explicit. Unless I missed something, as opposed to my current approach, testing in JS won't let me assert that an instrumented function follows the underlying steps of:
1. set the `trace_id` metadata
2. perform the instrumented operation
3. unset `trace_id` metadata
Simply because I don't have access to what happens under the hood (which I can do with the current relatively standard interfaces/mocks testing strategy).

I end up, however, asserting the higher level behavior of "performing instrumented operation X emits a `trace_id` metadata", and "performing instrumented operation Y right after emits a different `trace_id` value", which is more of an integration/e2e test, but with which I'm okay. Let me know if we have similar understandings and if that's satisfying from your perspective (or if I missed something 🤞🏻) 🙇🏻

mstoykov commented:
tl;dr I checked out the branch to test something and got carried away :( 


> perceived complexity introduced

The problem for me is that all the new changes exist to make it possible to test this (slightly) better. 

As you can see from my [try](https://github.com/grafana/k6/commit/64dd12e58432f7cc5bc899ffa74b2593c418f87a) at this the only thing I am not checking is the *concrete* value of the trace_id. 

I could be testing its format, but decided to not spend more time on it.

While I also worked on this I realized I also have an issue with the fact you are basically having a unit test, while I would argue that we need more of an integration test. One in which you do what the user will do and check that we do set the metadata correctly and then unset it after the operation. But doing this with actual JS code as the user will use it instead of what seems like a reimplementation of the actual code the user will call. 

Both have their place, but arguably we care more about the user experience instead of the underlying implementation details.

The "integration" test [here](https://github.com/grafana/k6/commit/5e5f4863d2424de191386419f346bb09a3203ae0) did take more fiddling in some parts due to the existence of two functions generating test setups, but mostly due to the requirements of the k6/http package so that it is *actually* functional. 

But even with that [both commits](https://github.com/grafana/k6/compare/experimental/tracing-test-metadata-wip?expand=1) are on par with just the code that is to add the Producers and add the mock ones. 

This also now lends itself to be expanded so we can add more tests directly in the package instead of adding it to actual integration tests https://github.com/grafana/k6/tree/master/cmd/tests

oleiade commented:
@mstoykov Thanks for the thorough examples, I see what you meant much better now indeed. I'm grateful for it 🙇🏻 

I hadn't consider the `instrumentCall` route to build an integration test, that's a good idea 👀 I'll integrate your approach in an integration test then indeed, and drop most of the interfaces/mocks I've added. 👍🏻

oleiade commented:
I've updated the PR according to your recommendations. I've integrated the tests you kindly put together Mihail, thanks again for that 🤝 I've also dropped the `Producer` abstraction.

codebien commented:
@oleiade the CI reported a panic

oleiade commented:
👍🏻 @codebien looking into it

codebien reviewed (approved):
LGTM. It seems you moved some functions, the next time would be helpful not to do it in the same commit you are doing the changes so it will be easier to review/skip.

mstoykov commented on js/modules/k6/experimental/tracing/trace_id.go:
```diff
@@ -76,7 +94,7 @@ func (t TraceID) Encode() (string, error) {
 	// is filled with random bytes.
 	randomness := make([]byte, 16-n)
-	err := binary.Read(rand.Reader, binary.BigEndian, randomness)
+	err := binary.Read(t.randSource, binary.BigEndian, randomness)
```
This is more of a general comment. But it seems strange that calling `TraceID.Encode` multiple times will get you *different* values. 

Like it seems to me this should just basically be a single function instead of having code that, creates a struct and then calls `Encode` on it as the only thing it does.

oleiade replied:
I've experimented around it. I have an idea of how to decouple trace id initialisation from encoding, but because the random part doesn't have a fixed size, it's gonna make things a bit more involved than I anticipated.

I'm already working on improving this in a separate PR. Resolving this for now, and stay tuned for the upcoming refactor PR 👍

mstoykov commented on js/modules/k6/experimental/tracing/client_test.go:
```diff
@@ -144,6 +149,105 @@ func TestClientInstrumentArguments(t *testing.T) {
+	callFn := func(args ...goja.Value) error {
+		gotMetadataTraceID, gotTraceIDKey := testCase.client.vu.State().Tags.GetCurrentValues().Metadata["trace_id"]
+		assert.True(t, gotTraceIDKey)
+		assert.NotEmpty(t, gotMetadataTraceID)
```
You went through the trouble of having this to be deterministic by setting a specific `randSource` and ... then you didn't test it is the specific value we expect :(

mstoykov commented on js/modules/k6/experimental/tracing/client_test.go:
```diff
@@ -144,6 +149,105 @@ func TestClientInstrumentArguments(t *testing.T) {
+	var sampleRead bool
+	for sampleContainer := range samples {
+		for _, sample := range sampleContainer.GetSamples() {
+			require.NotEmpty(t, sample.Metadata["trace_id"])
```
Here again you could check what the value is, given that it is deterministc, but you don't

mstoykov reviewed (changes requested):
I do like htat we now do not use the global rand source - that was definitel a problem I didn't notice before, but it is a concurrency concern as I mentioned in another PR.

But you then go and do not actually check the value :(.

codebien commented on js/modules/k6/experimental/tracing/trace_id.go:
```diff
@@ -45,6 +46,23 @@ type TraceID struct {
+// NewTraceID produces a new TraceID with the given prefix, code and time.
+//
+// It sets the traceID randomness source to the `rand.Reader` as a default.
+func NewTraceID(prefix int16, code int8, t time.Time, randSource io.Reader) TraceID {
```
```suggestion
func NewTraceID(prefix int16, code int8, t time.Time) TraceID {
```
It sounds not used

codebien replied:
I see, this is part of https://github.com/grafana/k6/pull/3019

oleiade commented:
@mstoykov done 👍🏻

mstoykov commented on js/modules/k6/experimental/tracing/client.go:
```diff
@@ -231,6 +194,61 @@ func (c *Client) Put(url goja.Value, args ...goja.Value) (*httpmodule.Response,
 	return c.Request(http.MethodPut, url, args...)
 }
 
+func (c *Client) instrumentedCall(call func(args ...goja.Value) error, args ...goja.Value) error {
+	if len(args) == 0 {
+		args = []goja.Value{goja.Null()}
+	}
+
+	traceContextHeader, encodedTraceID, err := c.generateTraceContext()
+	if err != nil {
+		return err
```
nit: this Pr would've been easier to code review if you didn't move this methods around in the codebase

oleiade replied:
Fair enough 🙇‍♂️Sorry for that. I'm trying to get better at isolating these kind of changes, but still having a hard time.
